kaldb/src/main/java/com/slack/kaldb/clusterManager/CacheSlotAssignmentService.java

```diff
@@ -0,0 +1,253 @@
+   * are preferred. No preference is given to specific cache slots, which may result in over/under
+   * utilization of specific cache slots.
+   *
+   * <p>If this method fails to successfully assign all the replicas needing slot assignment, the
+   * following iteration of this method would attempt to re-assign these until there are no more
+   * available replicas to assign.
+   *
+   * @return The count of successfully assigned cache slots
+   */
+  protected int assignCacheSlotsToReplicas() {
+    Timer.Sample assignmentTimer = Timer.start(meterRegistry);
+
+    List<CacheSlotMetadata> availableCacheSlots =
+        cacheSlotMetadataStore
+            .getCached()
+            .stream()
+            .filter(
+                cacheSlotMetadata ->
+                    cacheSlotMetadata.cacheSlotState.equals(
+                        Metadata.CacheSlotMetadata.CacheSlotState.FREE))
+            .collect(Collectors.toList());
+
+    List<String> assignedReplicaIds =
+        cacheSlotMetadataStore
+            .getCached()
+            .stream()
+            .map(cacheSlotMetadata -> cacheSlotMetadata.replicaId)
+            .filter(replicaId -> !replicaId.isEmpty())
+            .collect(Collectors.toList());
+
+    List<String> replicaIdsToAssign =
+        replicaMetadataStore
+            .getCached()
+            .stream()
+            // only replicas created in the last X mins
+            .filter(
+                replicaMetadata ->
+                    replicaMetadata.createdTimeUtc
+                        > Instant.now()
+                            .minus(
+                                managerConfig
+                                    .getReplicaEvictionServiceConfig()
+                                    .getReplicaLifespanMins(),
+                                ChronoUnit.MINUTES)
+                            .toEpochMilli())
+            // sort the list by the newest replicas first, in case we run out of available slots
+            .sorted(Comparator.comparingLong(ReplicaMetadata::getCreatedTimeUtc))
+            .map(replicaMetadata -> replicaMetadata.name)
+            .filter(replicaId -> !assignedReplicaIds.contains(replicaId))
+            .collect(Collectors.toList());
+
+    if (replicaIdsToAssign.size() > availableCacheSlots.size()) {
+      LOG.warn(
+          "Insufficient cache slots to assign replicas, wanted {} slots but had {} replicas",
+          replicaIdsToAssign.size(),
+          availableCacheSlots.size());
+      slotAssignInsufficientCapacity.increment(
+          replicaIdsToAssign.size() - availableCacheSlots.size());
+    } else if (replicaIdsToAssign.size() == 0) {
+      LOG.info("No replicas found requiring assignment");
+      assignmentTimer.stop(slotAssignTimer);
+      return 0;
+    }
+
+    AtomicInteger successCounter = new AtomicInteger(0);
+    List<ListenableFuture<?>> cacheSlotAssignments =
+        Streams.zip(
+                replicaIdsToAssign.stream(),
```

Comment:
Yeah, the behavior of discarding the remaining items is intentional. This follows the same pattern as the `RecoveryTaskAssignmentService` where prior to doing the Streams.zip we calculate if we can exit early or if we have too many replicas vs available slots. In the event we have too many replicas, we already log a warning with the delta between what was requested and what was available, as well as keep track of this with a dedicated meter.

Logging out the individual left-over items is likely tremendously valuable, as this would only be the replica ID (prefix + chunk timestamp + random UUID).